File: sbe/binance/spot_3_2/generated-python/spot_sbe/KlinesResponse.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple, Set
from io import BytesIO

class Klines:
    """Repeating group item."""

    def __init__(self):
        self.open_time = None
        self.open_price = None
        self.high_price = None
        self.low_price = None
        self.close_price = None
        self.volume = None
        self.close_time = None
        self.quote_volume = None
        self.num_trades = None
        self.taker_buy_base_volume = None
        self.taker_buy_quote_volume = None

class KlinesResponse:
    """SBE message: KlinesResponse."""

    TEMPLATE_ID = 203
    SCHEMA_ID = 3
    SCHEMA_VERSION = 2
    BLOCK_LENGTH = 2

    def __init__(self):
        self.price_exponent = None
        self.qty_exponent = None
        self.klines = []

    def _decode_klines_group(self, data: bytes, offset: int) -> Tuple[List[Klines], int]:
        """Decode repeating group."""
        pos = offset

        block_length = struct.unpack_from('<H', data, pos)[0]
        pos += 2
        num_in_group = struct.unpack_from('<I', data, pos)[0]
        pos += 4

        items = []
        for _ in range(num_in_group):
            item_start = pos
            item = Klines()

            item.open_time = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.open_price = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.high_price = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.low_price = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.close_price = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.volume = struct.unpack_from('<B', data, pos)[0]
            pos += 1
            item.close_time = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.quote_volume = struct.unpack_from('<B', data, pos)[0]
            pos += 1
            item.num_trades = struct.unpack_from('<q', data, pos)[0]
            pos += 8
            item.taker_buy_base_volume = struct.unpack_from('<B', data, pos)[0]
            pos += 1
            item.taker_buy_quote_volume = struct.unpack_from('<B', data, pos)[0]
            pos += 1

            # Skip to next block for forward compatibility
            pos = item_start + block_length
            items.append(item)

        return (items, pos)

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.price_exponent is not None:
            buffer.write(struct.pack('<b', self.price_exponent))
        if self.qty_exponent is not None:
            buffer.write(struct.pack('<b', self.qty_exponent))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        pos = 0

        self.price_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1
        self.qty_exponent = struct.unpack_from('<b', data, pos)[0]
        pos += 1

        # Skip to end of block for forward compatibility
        pos = 2

        self.klines, pos = self._decode_klines_group(data, pos)

File: sbe/binance/spot_3_1/generated-php/spot_sbe/KlinesResponse.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class Klines
{
    public int|float|null $openTime = null;
    public int|float|null $openPrice = null;
    public int|float|null $highPrice = null;
    public int|float|null $lowPrice = null;
    public int|float|null $closePrice = null;
    public int|float|null $volume = null;
    public int|float|null $closeTime = null;
    public int|float|null $quoteVolume = null;
    public int|float|null $numTrades = null;
    public int|float|null $takerBuyBaseVolume = null;
    public int|float|null $takerBuyQuoteVolume = null;
}

class KlinesResponse
{
    public const TEMPLATE_ID = 203;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 1;
    public const BLOCK_LENGTH = 2;

    public int|float|array|null $priceExponent = null;
    public int|float|array|null $qtyExponent = null;
    public array $klines = [];

    private function decodeKlinesGroup(string $data, int &$offset): array
    {
        $blockLength = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;
        $numInGroup = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;

        $items = [];
        for ($i = 0; $i < $numInGroup; $i++) {
            $itemStart = $offset;
            $item = new Klines();

            $item->openTime = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->openPrice = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->highPrice = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->lowPrice = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->closePrice = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->volume = unpack('C', substr($data, $offset, 1))[1];
            $offset += 1;
            $item->closeTime = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->quoteVolume = unpack('C', substr($data, $offset, 1))[1];
            $offset += 1;
            $item->numTrades = unpack('q', substr($data, $offset, 8))[1];
            $offset += 8;
            $item->takerBuyBaseVolume = unpack('C', substr($data, $offset, 1))[1];
            $offset += 1;
            $item->takerBuyQuoteVolume = unpack('C', substr($data, $offset, 1))[1];
            $offset += 1;

            // Skip to next block for forward compatibility
            $offset = $itemStart + $blockLength;
            $items[] = $item;
        }

        return $items;
    }

    public function encode(): string
    {
        $buffer = '';

        if ($this->priceExponent !== null) {
            $buffer .= pack('c', $this->priceExponent);
        }
        if ($this->qtyExponent !== null) {
            $buffer .= pack('c', $this->qtyExponent);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->priceExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->qtyExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;

        // Skip to end of block for forward compatibility
        $offset = 2;

        $this->klines = $this->decodeKlinesGroup($data, $offset);
    }
}
